### PR TITLE
Allows for hostile mobs to perform charge attacks, and adds a giant tackling lobster. 

### PR DESCRIPTION
**Pull request by ArcaneMusic**
## About The Pull Request
Sprites updated as of 8/10/2020, front sprite was all kryson, I did the side sprites.
image

Example video: https://youtu.be/EAVuPtJ_kFM

This adds a new mob, the lobstrosity. It's a upper-caves level icemoon mob, created as a genetic combination of a lobster, a crab, and a shrimp, let loose on the horrific Alaskan hellscape that is the icemoon. It's best feature, however, is that it makes use of a new feature of simplemobs, charging. Any hostile simplemob can be configured and customized in a uniform way in order to become a charger, giving it a basic leap attack, that will disable the user if they find themselves crashing into an unpassable turf, but otherwise lands a knockdown on the attacking target. Compensate of this, charging mobs will also gain a very noticable tell, where they shake a moment before they actually perform the charge to better signify that they're rearing up for a melee charge.
I'm not exactly sold on using `Shake()` for the visual tell, but I'm not quite sold on a visual to showcase that it's happening, so I'm looking for feedback on that specifically.
Additionally, there is a bit of weird movement that I've yet to really get ironed out first but for the most part it's completely functional.

After some feedback, there's now a reddish, lavaland-colored lobster, as well as an arctic themed, Alaskan lobster.

## Why It's Good For The Game

So on a massive tangential leap of logic, one thing ActionNinja brought up in a recent discord rant was that there are so many mechanical parts of SS13 combat that completely lack a tell whatsoever, or the tell is so quick you can blink and you'll miss it in a heartbeat. So I turned to the greatest enemy attack feedback mechanic that any game has ever seen: the Legend of Zelda Bull.

Flowchart below.
image

...also I've been looking for a tasteful way to get this sprite in-game for awhile now but wanted to really iron out a unique mechanic to go alongside it that would make it feel a little less bullshit to fight against.

## Changelog
:cl: ArcaneMusic + Kryson on Sprites
add: A migratory pack of Lobstrosities have appeared outside of icebox station, be advised of their charging attack.
code: Hostile simplemobs may now charge at targets.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm
Comment (ATH1909) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm:19]: this guy is really flimsy in comparison to other (upper level) icemoon mobs; look at wolves and polar bears for a comparison. also, he doesn't appear to regenerate health like the other upper level icemoon mobs do.

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm
Comment (ATH1909) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm:7]: If the lakes are underground, why do these guys only spawn on the top level? Shouldn't you be able to find these lakes on the lower levels (in, say, some form of special ruin with, like, several of these guys in it)? I mean, we have the blood drunk miner's ice lake, but you can't really intentionally put hostile simplemobs in a boss arena (that doesn't have hostile simplemob minions as its gimmick)...

**Comment by Iatots** [comment]
Should be plasma colored since I assume it lives in the plasma lakes.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm
Comment (Couls) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm:603]: Use the new cooldown system, COOLDOWN_CHECK, COOLDOWN_DECLARE etc etc

File: code/modules/mob/living/simple_animal/hostile/hostile.dm
Comment (ATH1909) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm:614]: so if any charging mob in the future is lavaproof, it'll lose its lava immunity after its first charge?

**Comment by CRITAWAKETS** [comment]
honestly could be good to give the charge attack to watchers to replace their mostly useless nowadays ice attack

**Comment by wesoda25** [comment]
Sprites should really reflect its environment. Looks completely out of place.

**Comment by ArcaneMusic** [comment]
Video added to the PR description with 4 directional sprites, and showcases charge activation.

**Comment by Whoneedspacee** [comment]
you might want to consider looking at how bubblegum does its charge tells or how wendigo shakes up and down

if you need help with understanding animations just dm me

**Comment by Tlaltecuhtli** [comment]
replace ice watchers with lobsters, they are ugly

**Comment by OverlyGeneric** [comment]
bullfighting gimmick is a go.

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm
Comment (ArcaneMusic) [reply code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm:7]: Mostly just because I wanted them to be a weaker mob, but not dilute the mob pool of the lower levels. Ya know, mid-level fodder.

**Comment by ArcaneMusic** [comment]
I'm not stale, you're stale, old, busted

**Comment by trerri** [comment]
nooooo why stale nooo

**Review by Twaticus — APPROVED** [review]

File: code/modules/mob/living/simple_animal/hostile/hostile.dm
Comment (ArcaneMusic) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm:287]: Diagnosing this now but I don't have the slighest clue why this is, so open to suggestions:
Charges are currently breaking because target is... itself. This target seems to work for the purposes of regular pathfinding, but when brought into the component it's causing the mob to throw itself at... itself.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm
Comment (ArcaneMusic) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm:515]: This feels slightly dirty, but I'd love to be told this is valid.

Comment (Rohesie) [reply code/modules/mob/living/simple_animal/hostile/hostile.dm:515]: The correct syntax, if I get what you are trying to do, would be this:
`if(!(SEND_SIGNAL(src, COMSIG_HOSTILE_POST_CHARGE) & COMPONENT_HOSTILE_POSTCHARGE_IMPACT))`
Or maybe this:
`if(SEND_SIGNAL(src, COMSIG_HOSTILE_POST_CHARGE) & (ALL & ~COMPONENT_HOSTILE_POSTCHARGE_IMPACT))`

**Comment by Rohesie** [comment]
@Qustinnus , feel like reviewing?

**Comment by Qustinnus** [comment]
@Rohesie I discussed the PR with Arcane in discord already. oranges disagreed because his opinions on components are different so it's probably good if you get a second opinion.

**Comment by Qustinnus** [comment]
But my general consensus was: This shouldn't be a component. It's not being used as one. Ideally we'd make behaviors datums and have an AI agent managing those. This is obviously way out of scope / unrealistic so in that case it's better of being in mobcode rather than being its own component, as it's not self-sufficient in any way and is basically being used as a proc holding datum rather than a component.

**Comment by Rohesie** [comment]
oranges didn't seem to disagree from what I saw
I agree on components needing to be self-sufficient, and not proc holders. That's not what they are.
What does Arcane think about this?

**Comment by Qustinnus** [comment]
oranges said it should be a component and use getcomponent. I'm not sure what arcane wants I think he wanted to keep it like this because oranges said it was fine, but I'm not really keen on it.

**Comment by Rohesie** [comment]
Yea, I don't like it either.
If we were to use `GetComponent` then it should be a datum held in a var instead.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm
Comment (TemporalOroboros) [reply code/modules/mob/living/simple_animal/hostile/hostile.dm:287]: I believe SEND_SIGNAL tacks the target (in this case src) to the beginning of the passed args. The datum relaying the signal will always be the first argument passed to any listeners. In this case specifically the args being passed are (src, target).

Comment (TemporalOroboros) [reply code/modules/mob/living/simple_animal/hostile/hostile.dm:287]: https://github.com/tgstation/tgstation/blob/b7d960b18727627287eb7ec2d8c0d712a4e8c3a7/code/__DEFINES/dcs/helpers.dm#L1-L5

**Comment by ArcaneMusic** [comment]
At this point? I'm honestly not sure, with this commit (Soon), it'll work functionally as a component to be added to any kind of hostile simplemob, but in the earlier commits I have a working version that works off the hostile mob's behavior tree.
I'm not sure what's preferable with what's available, but if I have to redo it for a 3rd time I'd like to know that there's a consensus on this.

**Comment by Qustinnus** [comment]
my opinion is "anything but a component"

**Comment by LemonInTheDark** [comment]
Status?

**Comment by Qustinnus** [comment]
it needs a maintainer to tell arcane what to actually do.

File: code/datums/components/charger.dm
Comment (LemonInTheDark) [inline code/datums/components/charger.dm:46]: I assume you missed this

**Comment by LemonInTheDark** [comment]
ahhhh that means someone needs to decide what to actually do. 

From what I understand the end goal is the make mob ai datum based, but that's wildly out of scope here.
So we're left with component behavior, which is currently acting as something like a proc holder. We still need to use booleans for defining behavior and all that.

There's also the mobcode version, which appears to be functionally the same without the airgap of a component, so it might even be slightly easier to read.

I would merge the mobcode, unless there's something I'm missing here it functions properly and doesn't detract from any functionality present or future. I don't think corrupting component space is a large concern, but in this case it seems to be needless.

**Comment by LemonInTheDark** [comment]
Actually you know what, lemme add an addendum to that. Like I said I don't really mind polluting component space, and since no other maint has said anything definite about it, I'd be fine with using getComponent() too. So long as stuff like charge_distance is moved fully to the component assignment, rather then being stored on the mob and passed to the component under the right circumstances.

You just gotta pick which system you want to use, rather then limboing between them.

**Comment by optimumtact** [comment]
make a datum/ai_behaviour class of items.

Implement datum/ai_behaviour/charge

give the mob a behaviour var that holds the current ai behaviour.

While you're there refactor all current behaviours into datums.

**Comment by optimumtact** [comment]
actually wait, I am not sure I stand at this position anymore

**Comment by optimumtact** [comment]
nevermind I argued myself out of any position, if you want to work with floyd then you can, if you don't then it's up to other maintainers to decide the future for signals.

**Comment by Whoneedspacee** [comment]
ideally all mobs should just be datum abilities that work on every mob so we don’t have to copy paste shit.

bubblegum has a charge, leaper has a charge, etc.

there’s an argument against components because this is something that would be entirely mob based and never really be used on anything else.

an upside of them being datums as well is they can be tied into our current ability system like i already do with the megafauna abilities and that selection could just be built in, not that you couldn’t do this with components, but tying this into our existing abilities system makes sense.

turning this into a mob ability thats a targeted click like the wizard spells would be cool too since it would have great precision, and would also leave the charge ability to be used by carbon mobs as well.

if this is something you could do that’d be great, but to be honest a component isnt that bad, and who knows if someone will be fucked to rework all mob abilities.

**Comment by Qustinnus** [comment]
im planning to because working with simplemobs takes up more time than it would to just re-work it to be less shit.

**Comment by Whoneedspacee** [comment]
then yeah this should probably be engrained into our current abilities system like wizard spells are, and be callable from the mobcode.

this would make player control of the mob auto complete as well, no extra code required.

File: code/datums/components/charger.dm
Comment (Jared-Fogle) [inline code/datums/components/charger.dm:30]: All signal procs need SIGNAL_HANDLER.

File: code/datums/components/charger.dm
Comment (Jared-Fogle) [reply code/datums/components/charger.dm:46]: If this is intentional, please move this to something like `addtimer` so the signal stays non-blocking.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm
Comment (AnturK) [reply code/modules/mob/living/simple_animal/hostile/hostile.dm:515]: Shouldn't all this be in the charger component.  This structure doesn't make much sense, you have very specific signal here that won't be used for anything else.
Any reason it can't be implemented as something like:
throw_impact > Sends on impact signal > Charger Component blocks default action, does it stuff, ideally calling some on impact signal other stuff can subscribe to

File: code/modules/mob/living/simple_animal/hostile/hostile.dm
Comment (AnturK) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm:287]: I'd name this signal something else like try ON_TARGET_APPROACH or similar.

Comment (Qustinnus) [reply code/modules/mob/living/simple_animal/hostile/hostile.dm:287]: image

File: code/modules/mob/living/simple_animal/hostile/hostile.dm
Comment (LemonInTheDark) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm:606]: This should be a timer, since this is done in npc_pool.process() right?

File: code/modules/mob/living/simple_animal/hostile/hostile.dm
Comment (LemonInTheDark) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm:633]: ditto? maybe?

File: code/modules/mob/living/simple_animal/hostile/hostile.dm
Comment (LemonInTheDark) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm:639]: Have you thought about using the other args of this proc?

**Pull request merged**
